sway_libs/src/fixed_point/ufp/ufp128/ufp128.sw

```diff
@@ -0,0 +1,234 @@
+        let res_u256 = self_multiply_other >> 64;
+        if res_u256.a != 0 || res_u256.b != 0 {
+            // panic on overflow
+            revert(0);
+        }
+        Self::from((res_u256.c, res_u256.d))
+    }
+}
+
+impl core::ops::Divide for UFP128 {
+    /// Divide a UFP128 by a UFP128. Panics if divisor is zero.
+    fn divide(self, divisor: Self) -> Self {
+        let mut s = self;
+        let zero = UFP128::zero();
+        let u128_max = U128::max();
+
+        assert(divisor != zero);
+
+        // Conversion to U128 done to ensure no overflow happen
+        // and maximal precision is avaliable
+        // as it makes possible to multiply by the denominator in 
+        // all cases
+        let self_u256 = U256::from((0, 0, self.value.upper, self.value.lower));
+        let divisor_u256 = U256::from((0, 0, divisor.value.upper, divisor.value.lower));
+
+        // Multiply by denominator to ensure accuracy 
+        let res_u256 = (self_u256 << 64) / divisor_u256;
+
+        if res_u256.a != 0 || res_u256.b != 0 {
+            // panic on overflow
+            revert(0);
+        }
+        Self::from((res_u256.c, res_u256.d))
+    }
+}
+
+impl UFP128 {
+    pub fn recip(number: UFP128) -> Self {
+        let one = UFP128::from((1, 0));
+
+        one / number
+    }
+
+    pub fn floor(self) -> Self {
+        Self::from((self.value.upper, 0))
+    }
+
+    pub fn ceil(self) -> Self {
+        let val = self.value;
+        if val.lower == 0 {
+            return Self::from((val.upper, 0));
+        } else {
+            return Self::from((val.upper + 1, 0));
+        }
+    }
+}
+
+impl UFP128 {
+    pub fn round(self) -> Self {
+        let floor = self.floor();
+        let ceil = self.ceil();
+        let diff_self_floor = self - floor;
+        let diff_ceil_self = ceil - self;
+        if diff_self_floor < diff_ceil_self {
+            return floor;
+        } else {
+            return ceil;
+        }
+    }
+
+    pub fn trunc(self) -> Self {
+        Self::from((self.value.upper, 0))
+    }
+
+    pub fn fract(self) -> Self {
+        Self::from((0, self.value.lower))
+    }
+}
+
+impl Root for UFP128 {
+    fn sqrt(self) -> Self {
+        let nominator_root = self.value.sqrt();
+        let nominator = nominator_root * U128::from((0, 2 << 32));
+        Self::from((nominator.upper, nominator.lower))
+    }
+}
+
+impl Exponentiate for UFP128 {
+    fn pow(self, exponent: Self) -> Self {
+        let nominator_pow = self.value.pow(exponent.value);
+        let u128_1 = U128::from((0, 1));
+        let u128_2 = U128::from((0, 2));
+        let u128_64 = U128::from((0, 64));
+        let two_pow_64_n_minus_1 = u128_2.pow(u128_64 * (exponent.value - u128_1));
+        let nominator = nominator_pow / two_pow_64_n_minus_1;
+        Self::from((nominator.upper, nominator.lower))
+    }
+}
+
+// TODO: uncomment and change accordingly, when signed integers will be added
```

Comment:
Signed integers are good to go, no?

Reply:
log function should changed, so this part has to stay as is, for now